src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -315,6 +320,12 @@ public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, V
                 attributeIcons: attributeImages);
 
             item.Properties.AddProperty(RoslynItem, roslynItem);
+
+            if (roslynItem.UseEditorCompletionItemCache)
```

Comment:
hmmm.. so we have a circular reference. 

I guess it is not using CompletionItem.Properties since it is immutable for us?

Reply:
I see CompletionItem has a bunch of this mutable internal states. won't it better just have mutable InternalProperties or something rather than keep adding this one off mutable state on CompletionItem?

Reply:
I believecircular reference is fine in terms of GC? Would it cause any trouble? Using the property bag will incur another allocation of immutableHashSet, which I'd like to avoid if possible given that we might create a large amount of those completion items on one shot.

Reply:
circular ref should be no problem. :) And we def want to avoid the allocs of the dictionaries as well.  ComplationItem.Properties is also only for string->string mappings.  this is an object-instantiation cache.